gateway-service-impl/src/main/java/org/hypertrace/gateway/service/entity/query/ExecutionTreeBuilder.java

```diff
@@ -305,6 +318,106 @@ QueryNode buildFilterTree(EntitiesRequest entitiesRequest, Filter filter) {
     }
   }
 
+  // filters and order by on QS, but you can still have selection on EDS
+  QueryNode buildAndFilterTree(EntitiesRequest entitiesRequest) {
+    // If the filter by and order by are from QS (and selections are on other sources), pagination
+    // can be pushed down to QS
+    // Since the filter and order by are from QS, there won't be any filter on other
+    // sources
+    if (sourceSetsIfFilterAndOrderByAreFromSameSourceSets.contains(QS.name())) {
+      executionContext.setSortAndPaginationNodeAdded(true);
+      return createQsDataFetcherNodeWithLimitAndOffset(entitiesRequest);
+    }
+
+    Map<AttributeSource, Filter> sourceToAndFilterMap =
+        new EnumMap<>(buildSourceToAndFilterMap(entitiesRequest.getFilter()));
+
+    // qs node as the pivot node to fetch time range data
```

Comment:
So, there won't be any query w/o time range filter, right?

Reply:
right